co-noir/ultrahonk/src/parse/proving_key.rs

```diff
@@ -0,0 +1,359 @@
+                let next_cycle_node_index = if node_idx == copy_cycle_size - 1 {
+                    0
+                } else {
+                    node_idx + 1
+                };
+                let next_cycle_node = &copy_cycle[next_cycle_node_index];
+                let current_row = current_cycle_node.gate_index as usize;
+                let next_row = next_cycle_node.gate_index;
+
+                let current_column = current_cycle_node.wire_index as usize;
+                let next_column = next_cycle_node.wire_index;
+                // Point current node to the next node
+                mapping.sigmas[current_column][current_row].row_index = next_row;
+                mapping.sigmas[current_column][current_row].column_index = next_column;
+
+                let first_node = node_idx == 0;
+                let last_node = next_cycle_node_index == 0;
+
+                if first_node {
+                    mapping.ids[current_column][current_row].is_tag = true;
+                    mapping.ids[current_column][current_row].row_index =
+                        real_variable_tags[cycle_index];
+                }
+                if last_node {
+                    mapping.sigmas[current_column][current_row].is_tag = true;
+
+                    // TODO(Zac): yikes, std::maps (tau) are expensive. Can we find a way to get rid of this?
+                    mapping.sigmas[current_column][current_row].row_index = *circuit_constructor
+                        .tau
+                        .get(&real_variable_tags[cycle_index])
+                        .expect("tau must be present  ");
+                }
+            }
+        }
+
+        // Add information about public inputs so that the cycles can be altered later; See the construction of the
+        // permutation polynomials for details.
+        let num_public_inputs = circuit_constructor.public_inputs.len();
+
+        for i in 0..num_public_inputs {
+            let idx = i + pub_inputs_offset;
+            mapping.sigmas[0][idx].row_index = idx as u32;
+            mapping.sigmas[0][idx].column_index = 0;
+            mapping.sigmas[0][idx].is_public_input = true;
+            if mapping.sigmas[0][idx].is_tag {
+                tracing::warn!("MAPPING IS BOTH A TAG AND A PUBLIC INPUT");
+            }
+        }
+
+        mapping
+    }
+
+    fn compute_honk_style_permutation_lagrange_polynomials_from_mapping(
+        permutation_polynomials: &mut [Polynomial<P::ScalarField>],
+        permutation_mappings: Mapping,
+        circuit_size: usize,
+    ) {
+        let num_gates = circuit_size;
+
+        for (current_permutation_mappings, current_permutation_poly) in permutation_mappings
+            .into_iter()
+            .zip(permutation_polynomials)
+        {
+            debug_assert_eq!(current_permutation_mappings.len(), num_gates);
+            debug_assert_eq!(current_permutation_poly.len(), num_gates);
+
+            // TODO Barrettenberg uses multithreading here
```

Comment:
why can't we, seems pretty `par_iter`able?

Reply:
I ignored multithreading completely so far, Let's do all multithreading after the merge